### ISSUE TITLE
Add new `collect` query clause for doing aggregation

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#1137]
This is part of #441.

**Comment by jclark** [comment]
Concept is as follows from https://github.com/ballerina-platform/ballerina-spec/issues/441#issuecomment-1046406514:

- there is a new clause that is an alternative to `select`
- syntax for new clause uses `=>` rather than a new keyword
- the  body of new clause would initially be limited to a call expression (method or function call)
- variables bound in the query pipeline can occur only in the last argument to the call expression
- an unqualified function name is mapped onto a langlib function using the argument types
- a method call or qualified function name is resolved in the usual way; the function must be either varargs or accept a stream argument

So a simple example would look like:

```
var total = from var { cost } in orders => sum(cost);
```

More complex example:
```
string nameList =
  from var { firstName, lastName } in persons
  => ", ".join(firstName + " " + lastName);
```

From a user perspective, effect is that the last argument is repeated once for each variable binding.

**Comment by sanjiva** [comment]
I'm not sure I like the use of `=>` here as it will look a bit off in an expression bodied function:

```
function foo (Person p) returns Student =>
   {
      name: p.name,
      totalFee: from var { cost } in p.something => sum(cost)
   };

IMO a keyword will look nicer.

**Comment by sanjiva** [comment]
Also how does this work:
```an unqualified function name is mapped onto a langlib function using the argument types```
Does that mean any function I call MUST be in another module? We don't allow `self:foo()` IIRC.

**Comment by jclark** [comment]
The semantics of this can be explained in terms of the _implicitly spread_ variable concept described in https://github.com/ballerina-platform/ballerina-spec/issues/1134#issuecomment-1174577892. The effect of `=>` is to combine all incoming frame into a single group without any grouping variables, thus making all variables bound in preceding query clauses into implicitly spread clauses.

**Comment by jclark** [comment]
> Does that mean any function I call MUST be in another module? 

This is just for aggregations. This means a user doesn't get to define their own kinds of aggregation in the same module. User-defined aggregations are a pretty niche thing. Almost all the time people will be using the aggregations we provide (min, max, avg, sum, count etc).

**Comment by jclark** [comment]
> I'm not sure I like the use of => here as it will look a bit off in an expression bodied function

I take your point about expression-bodied functions.

Difficulty with new keyword is:

- coming up with a good keyword (there's nothing in SQL to mimic)
- avoid incompatibility issues from introduction of a new keyword (though perhaps we can solve them in #1136)

The other idea I had was `select...` (the idea being you are doing one `select` with all the values, rather than one for each), which avoids incompatibilities with a new keyword. Effectively what `select...` is doing here

```
var total = from var { cost } in orders select... sum(cost);
```

is creating a variable cList that is a list containing one member for each value of `cost` and then evaluating `decimal:sum(...cList)`.

Alternatively something completely new:

- `aggregate`
- `combine`

**Comment by sanjiva** [comment]
In Siddhi we used `aggregate`: https://siddhi.io/en/v4.x/docs/query-guide/#incremental-aggregation.

**Comment by jclark** [comment]
Siddhi's use of `aggregate` is very different from what we need a keyword for. Siddhi's `aggregate by` is like `group by` but for combining things over time periods: it is used in conjunction with `select` (like `group by`) rather than instead of it. I suspect we will need a clause with similar semantics to the Siddhi `aggregate` clause when we do #440.

**Comment by jclark** [comment]
I suggest `collect`.

**Comment by KavinduZoysa** [comment]
@jclark, in the previous example, it is mentioned that the two examples are effectively the same. But I think the type of output `total` is not the same. 
In the second example, output `total` should be a single element array where the value of that element equals the `sum(cost)`. But in the first example, the output `total` contains `sum(cost)`. 
Is that correct?

**Comment by jclark** [comment]
@KavinduZoysa Yes, you are correct.

**Comment by jclark** [comment]
There's another difference  between `group by/select` and `collect` as regards empty group. A `group by` clause never creates an empty group, but with `collect` there is the potential that the incoming stream of frames is empty. So we need to figure out how to deal with that.  Some aggregation functions can deal with that naturally (e.g. `sum`); for others, it is less straightforward.

**Comment by jclark** [comment]
@gimantha You were going to make a comment here...

**Comment by gimantha** [comment]
> @gimantha You were going to make a comment here...
I am sorry for the delay. 
Can a query expression have the `collect` clause after a  `group by` clause? It will be like having two group by clauses together where the second group by clause will take all the grouped frames coming from first group by clause and create one frame with them.

**Comment by gimantha** [comment]
> @gimantha You were going to make a comment here...

I am sorry for the delay. 
Can a query expression have the `collect` clause after a  `group by` clause? It will be like having two group by clauses together where the second group by clause will take all the grouped frames coming from first group by clause and create one frame with them.

**Comment by jclark** [comment]
@gimantha As you say, this is really the same problem as whether we can have multiple `group by` clauses, which needs to be addressed in #1134.

**Comment by jclark** [comment]
Fixed by [5e76414](https://github.com/ballerina-platform/ballerina-spec/commit/5e764141a5bbad3e8162b67434fd8228e9b1c314).
